text/3829-de-rfc-unsized-locals.md

```diff
@@ -0,0 +1,146 @@
+- Feature Name: `unsized_locals`
+- Start Date: 2025-06-02
+- RFC PR: [rust-lang/rfcs#3829](https://github.com/rust-lang/rfcs/pull/3829)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+_Following the great success of the [type ascription de-RFC], we now present the next one in the series!_
+
+# Summary
+[summary]: #summary
+
+Unsized locals ([RFC 1909], called "unsized rvalues" originally)
+has been a merged RFC for eight years with no clear path to stabilization.
+
+There is still a very large gap in the implementation in rustc that hasn't been addressed for many years and there are several language problems with this feature. 
+
+This RFC intends to advocate for the feature being removed entirely with a fresh RFC being necessary to add a similiar feature again.
+
+# Demotivation
+[demotivation]: #demotivation
+
+The `unsized_locals` feature is simple to explain on the surface: local variables no longer have to be `Sized`.
+
+```rust
+#![feature(unsized_locals)]
+fn main() {
+    let x = *Box::from("hello, world!");
+}
+```
+
+This will dynamically allocate space on the stack for the string.
+C has a similar feature, [`alloca`] and variable length arrays (VLA), the latter having been made optional to implement in C11.
+
+## Lack of Proper Implementation
+
+This feature has never been properly implemented in rustc.
+The variable length array form proposed in the RFC still doesn't exist at all, only the unsized local variable does.
+It is implemented in the type checker and codegen, but lacking MIR semantics and therefore unimplemented in the compile time function evaluator.
+This is very significant, as MIR semantics govern how the feature should behave precisely in the first place.
+Without them, they cannot work in `const` and optimizations are likely broken around them.
```

Comment:
I would like to understand the “MIR semantics” issues in more detail. What exactly would the compiler team need to see before it would become possible to move forward? https://github.com/rust-lang/compiler-team/issues/630 says:

> The [MIR semantics] that const eval used to implement - before support was removed - were known to be awkward and surprising.

But doesn’t link to any of that discussion. It’s hard for people to step up and address the problems, without a clear summary of what they are.

Reply:
Even then, a detailed summary of the problems would also have to be updated for changes in MIR semantics since then, no? Which may be potentially quite a bit.

Reply:
I found https://github.com/rust-lang/rust/issues/48055#issuecomment-1837424794 which seems to be what I was looking for.

Reply:
I added a link to the comment in 5bc11846302f5fe76326f065c33b554339979e91

Reply:
I started a discussion about fixing the MIR on Zulip: https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/MIR.20semantics.20for.20unsized.20locals